### PR TITLE
feat(ecs): secret JSON key for environment variables

### LINKED ISSUE #5665: How to use only password of secret from RDS in ECS
## :question: General Issue

I wonder what's the proposed way to use the generated secret from an rds instance/cluster.

```typescript
const database = new rds.DatabaseCluster(this, 'DatabaseCluster', {
            removalPolicy: RemovalPolicy.DESTROY,
            defaultDatabaseName: scope.getConfig(this, 'wordpressDbName'),
            masterUser: {
                username: 'wordpress',
            },
            instances: 1,
            engine: rds.DatabaseClusterEngine.AURORA,
            instanceProps: {
                instanceType: ec2.InstanceType.of(ec2.InstanceClass.T3, ec2.InstanceSize.SMALL),
                vpc: vpc,
            }
        })
```

In ECS I would like to get only the password. All other information is known anyway.

```typescript
WORDPRESS_DB_PASSWORD: ecs.Secret.fromSecretsManager(database.secret?.secretValueFromJson('password')),
```

but this does not work. Modifications inside the container should not be the solution.

### Environment

  - **CDK CLI Version:** 1.19.0
  - **Module Version:** 1.19.0

### PR DESCRIPTION
**Pull request by jogold**
### Commit message

feat(ecs): secret JSON key for environment variables

Amazon Elastic Container Service now supports reading AWS Secrets Manager secrets from a key within a JSON object.

See https://aws.amazon.com/about-aws/whats-new/2020/02/amazon-ecs-now-supports-aws-secrets-manager-version-and-json-keys/

Closes #5665

### End of commit message

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-ecs/lib/container-definition.ts [hunk 1]
```diff
@@ -26,10 +26,16 @@ export abstract class Secret {
...
    * Manager.
+   *
+   * @param secret the secret stored in AWS Secrets Manager
+   * @param key the name of the key in a key-value pair with the value that you
+   * want to set as the environment variable value.  Only values in JSON format
+   * are supported. If you do not specify a JSON key, then the full contents of
+   * the secret is used.
    */
-  public static fromSecretsManager(secret: secretsmanager.ISecret): Secret {
+  public static fromSecretsManager(secret: secretsmanager.ISecret, key?: string): Secret {
```
Comment (eladb) [inline packages/@aws-cdk/aws-ecs/lib/container-definition.ts]: `key` => `field`?

**Comment by jogold** [comment]
Any update/feedback here?

**Comment by alukach** [comment]
This is looking great and would be very useful. @eladb any thoughts on ability to get this merged?

File: packages/@aws-cdk/aws-ecs/test/fargate/integ.secret-json-key.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+import * as cdk from '@aws-cdk/core';
+import * as ecs from '../../lib';
+
+const app = new cdk.App();
+const stack = new cdk.Stack(app, 'aws-ecs-integ-secret-json-key');
+
+const secret = new secretsmanager.Secret(stack, 'Secret', {
+  generateSecretString: {
+    generateStringKey: 'password',
+    secretStringTemplate: JSON.stringify({ username: 'user '})
```
Comment (sonofachamp) [inline packages/@aws-cdk/aws-ecs/test/fargate/integ.secret-json-key.ts:11]: If the extra space in `user ` is unnecessary, can you please remove it?

File: packages/@aws-cdk/aws-ecs/test/fargate/integ.secret-json-key.expected.json [hunk 3]
```diff
@@ -0,0 +1,117 @@
+{
+  "Resources": {
+    "SecretA720EF05": {
+      "Type": "AWS::SecretsManager::Secret",
+      "Properties": {
+        "GenerateSecretString": {
+          "GenerateStringKey": "password",
+          "SecretStringTemplate": "{\"username\":\"user \"}"
```
Comment (sonofachamp) [inline packages/@aws-cdk/aws-ecs/test/fargate/integ.secret-json-key.expected.json:8]: If the extra space in `user ` is unnecessary, can you please remove it?

File: packages/@aws-cdk/aws-ecs/lib/container-definition.ts [hunk 4]
```diff
@@ -26,10 +26,16 @@ export abstract class Secret {
   /**
    * Creates a environment variable value from a secret stored in AWS Secrets
    * Manager.
+   *
+   * @param secret the secret stored in AWS Secrets Manager
+   * @param field the name of the field with the value that you want to set as
+   * the environment variable value.  Only values in JSON format are supported.
```
Comment (sonofachamp) [inline packages/@aws-cdk/aws-ecs/lib/container-definition.ts:32]: Please remove the extra space between sentences in doc string.

**Comment by RikJansen81** [comment]
It seems this cloudformation-feature is only supported for ECS using the EC2 launch type, but not for Fargate:

https://github.com/aws/containers-roadmap/issues/385#issuecomment-591634935
https://docs.aws.amazon.com/AmazonECS/latest/developerguide/specifying-sensitive-data-secrets.html

Perhaps an idea to add this to the docs (and review the fargate-tests in the https://github.com/aws/aws-cdk/commit/97959f6ba40a4a576fc914772206623900d72add)?

**Comment by jogold** [comment]
> It seems this cloudformation-feature is only supported for ECS using the EC2 launch type, but not for Fargate:
> 
> [aws/containers-roadmap#385 (comment)](https://github.com/aws/containers-roadmap/issues/385#issuecomment-591634935)
> [docs.aws.amazon.com/AmazonECS/latest/developerguide/specifying-sensitive-data-secrets.html](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/specifying-sensitive-data-secrets.html)
> 
> Perhaps an idea to add this to the docs (and review the fargate-tests in the [97959f6](https://github.com/aws/aws-cdk/commit/97959f6ba40a4a576fc914772206623900d72add))?

You are right, see #7317

**Comment by ivawzh** [comment]
Awesome work! @jogold May I plz confirm if `ecs.Secret.fromSecretsManager` is going to support RDS secret autorotation?

**Comment by jogold** [comment]
> Awesome work! @jogold May I plz confirm if `ecs.Secret.fromSecretsManager` is going to support RDS secret autorotation?

What do you mean exactly?

**Comment by ivawzh** [comment]
I guess I have misunderstood the rotation feature of Secrets Manager. I thought it automatically triggers my services to update the password and redeploy the service when a secret is rotated. It seems not the case.

**Comment by picosam** [comment]
> I guess I have misunderstood the rotation feature of Secrets Manager. I thought it automatically triggers my services to update the password and redeploy the service when a secret is rotated. It seems not the case.

Hello, that's what I thought as well. May you explain what the case is as you found it out to be?

**Pull request merged**